webapp/app.py

```diff
@@ -1,25 +1,108 @@
-from flask import Flask, jsonify, redirect, render_template, url_for
+import logging
+
+from flask import Flask, Request, jsonify, redirect, render_template, session, url_for
+from flask_login import (
+    LoginManager,
+    UserMixin,
+    login_required,
+    logout_user,
+)
 from werkzeug.wrappers.response import Response
 
+from webapp.config import Config
 from webapp.exceptions import ECSTaskLogStreamDoesNotExistError
-from webapp.utils import get_task_status_and_logs
+from webapp.utils import get_task_status_and_logs, parse_oidc_data
 from webapp.utils.aws import ECSClient
 
+logger = logging.getLogger(__name__)
+
+
+class User(UserMixin):
+    """Class representing users of the app.
+
+    This class stores credentials (MIT ID, name, and email)
+    about logged in users.
+    """
+
+    def __init__(
+        self, id: str, name: str, email: str | None = None  # noqa: A002
+    ) -> None:
+        self.id = id
+        self.name = name
+        self.email = email
+
 
 def create_app() -> Flask:
     app = Flask(__name__)
+    app.secret_key = Config().SECRET_KEY
+
+    login_manager = LoginManager()
+    login_manager.init_app(app)
+
+    @login_manager.request_loader
+    def load_user_from_request(request: Request) -> User | None:
+        """Callback to login users using data from request headers.
+
+        For this application, login is the process of identifying the user
+        by decoding data provided in the request headers. This method
+        will:
+
+        * Retrieve OpenID Connect (OIDC) access tokens and user data from request headers.
+           * If access tokens or user data are missing from the request headers,
+             user cannot login.
+        * Stores active access tokens and user data in the Flask session.
+        * Create or refresh a User instance with parsed user data from OIDC.
+
+        An expired access token is marked by a mismatch between the OIDC
+        access token retrieved from the request header and the access token
+        currently stored in the Flask session. If there is a mismatch,
+        the method will parse the updated OIDC data in the request header
+        and refresh the User instance with the updated user data.
+        """
```

Comment:
Great docstring!  Think this is a conceptually complex topic, and I feel like this docstring really succinctly captures the important bits.